### PR TITLE
LLT-5531: Directory per test for logs

### PR DESCRIPTION
**Pull request by tomasz-grz**
### Problem
Our nat-lab test log files are getting so long it sometimes manages to reach filename limit’s in Windows.
And it’s extremely hard to focus in on all of data produced for a singular tests. 

### Solution
Instead of storing all logging artifacts in one directory, with long names. Create a sub directories
for each test, that stores log for all logs/pcaps etc created for this test.

### :ballot_box_with_check: Definition of Done checklist
- [x] Commit history is clean ([requirements](../blob/main/docs/git_commit_messages_requirements.md))
- [x] README.md is updated
- [x] Functionality is covered by unit or integration tests

### CODE REVIEW

**Review by gytsto — DISMISSED** [review]
+1 great work

File: nat-lab/tests/utils/testing.py [hunk 1]
```diff
@@ -10,8 +12,102 @@ def unpack_optional(opt: Optional[T]) -> T:
...
+    print(datetime.now(), "PYTEST_CURRENT_TEST is None")
+    return None
+
+
+def format_path_string(input_str: str) -> str:
+    """
+    Format the input string to be safely used as a system path or file name.
+    """
+    # truncate if it's longer than 64 characters to limit long absolute paths
+    # replace any non-alphanumeric characters, and the last "_" if it is present
```
Comment (jjanowsk) [inline nat-lab/tests/utils/testing.py:88]:
```
# truncate if it's longer than 64 characters
# replace any non-alphanumeric characters with "_", 
# remove trailing "_"
```

File: nat-lab/tests/utils/testing.py [hunk 2]
```diff
@@ -10,8 +12,102 @@ def unpack_optional(opt: Optional[T]) -> T:
...
+    """
+    Returns a string representing the case name of the currently running test.
+
+    The result is derived from PYTEST_CURRENT_TEST env variable.
+
+    For example:
+    "tests/test_mesh_plus_vpn.py::test_vpn_plus_mesh_over_direct" -> "test_vpn_plus_mesh_over_direct"
+    """
     test_name = os.environ.get("PYTEST_CURRENT_TEST")
     if test_name is not None:
```
Comment (jjanowsk) [inline nat-lab/tests/utils/testing.py:47]: if test_name:

File: nat-lab/tests/utils/testing.py [hunk 3]
```diff
@@ -10,8 +12,102 @@ def unpack_optional(opt: Optional[T]) -> T:
...
-        return "".join([x if x.isalnum() else "_" for x in test_name.split(" ")[0]])
-    return test_name
+        try:
+            test_part = test_name.split("::", 1)[-1].split(" ")[0].split("[", 1)[0]
+            return format_path_string(test_part)
+        except IndexError:
```
Comment (jjanowsk) [inline nat-lab/tests/utils/testing.py:51]: Personally I would not try to catch it and I would let it fail so we know we have some edge cases to cover early

**Review by jjanowsk — COMMENTED** [review]
Leaving some nitpicks to consider

File: nat-lab/tests/utils/testing.py [hunk 4]
```diff
@@ -10,8 +12,102 @@ def unpack_optional(opt: Optional[T]) -> T:
...
+
+def get_current_test_log_path(base_dir: str) -> str:
+    """
+    Returns a path to the log files of the currently running test.
+    """
+    # format the base_dir
+    if base_dir == "":
+        base_dir = "logs"
+    else:
+        base_dir = format_path_string(base_dir)
```
[the remark this answers is not in the document]
Comment (tomasz-grz) [reply nat-lab/tests/utils/testing.py:100]: Good idea, I totally agree, but then it can not be re-used for `coredumps` for example in case we decide to use the same structure.. but then we can just put it together in the logs folder 🤔

File: nat-lab/tests/utils/testing.py [hunk 5]
```diff
@@ -10,8 +12,102 @@ def unpack_optional(opt: Optional[T]) -> T:
     return opt
 
 
-def test_name_safe_for_file_name():
+def get_current_test_file_name() -> str | None:
```
[the remark this answers is not in the document]
Comment (tomasz-grz) [reply nat-lab/tests/utils/testing.py:15]: But would this ever be useful in the future in case we want to change how the logs are structured again? If not then I will remove.

File: nat-lab/tests/interderp_cli.py [hunk 6]
```diff
@@ -52,7 +53,7 @@ async def save_logs(self, count) -> None:
             await process.execute()
             container_id = process.get_stdout().strip()
         else:
-            container_id = str(self._connection.target_os)
+            container_id = format_path_string(str(self._connection.target_os))
```
[the remark this answers is not in the document]
Comment (tomasz-grz) [reply nat-lab/tests/interderp_cli.py:56]: `self._connection.target_os` has the form of `TargetOS.Windows` or `TargetOS.Mac` resulting in the files being named `TargetOS.Windows.log` which doesn't follow the rest of the naming convention.

[the remark this answers is not in the document]
Comment (tomasz-grz) [reply nat-lab/tests/interderp_cli.py:56]: On the other hand, could I do something like `self._connection.target_os.value` to get rid of the `TargetOS` part?

File: nat-lab/tests/utils/testing.py (hunk 2 above)
Comment (tomasz-grz) [reply nat-lab/tests/utils/testing.py:47]: good `catch` ;D

File: nat-lab/tests/utils/testing.py (hunk 3 above)
Comment (tomasz-grz) [reply nat-lab/tests/utils/testing.py:51]: But do we want to fail the whole pipeline in case the logs are not formatted correctly?

File: nat-lab/tests/utils/testing.py [hunk 7]
```diff
@@ -10,8 +12,102 @@ def unpack_optional(opt: Optional[T]) -> T:
...
+            file_part = test_name.split("::")[0].split(".")[0].split("/")[-1]
+            return format_path_string(file_part)
+        except IndexError:
+            print(datetime.now(), "get_current_test_file_name() IndexError", test_name)
+            return test_name
+    print(datetime.now(), "PYTEST_CURRENT_TEST is None")
+    return None
+
+
+def get_current_test_case_name() -> str | None:
```
[the remark this answers is not in the document]
Comment (tomasz-grz) [reply nat-lab/tests/utils/testing.py:37]: I guess it would be more like `get_current_test_case_name_and_parameters() -> (str, str | None) | None` since some tests don't have parameters, though I like the idea of avoiding code duplication.

File: nat-lab/tests/interderp_cli.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tomasz-grz) [reply nat-lab/tests/interderp_cli.py:56]: no `target_os.value` just returns an integer 😿

File: nat-lab/tests/utils/testing.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jjanowsk) [reply nat-lab/tests/utils/testing.py:100]: I'm usually a fan of a strategy: add it when it's actually needed

File: nat-lab/tests/utils/testing.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jjanowsk) [reply nat-lab/tests/utils/testing.py:15]: Then it can be easily added. Same as above

File: nat-lab/tests/utils/testing.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jjanowsk) [reply nat-lab/tests/utils/testing.py:37]: I missed this but I think the signature can stay and empty string can be returned as param in such case. This way when you glue it togeter you can get rid of one if statement.

File: nat-lab/tests/utils/testing.py (hunk 3 above)
Comment (jjanowsk) [reply nat-lab/tests/utils/testing.py:51]: Why not. It will literally never happen. Unless in 5 years someone in pytest goes crazy and changes this format. Then we want to catch it and fix this place :)

Comment (tomasz-grz) [reply nat-lab/tests/utils/testing.py]: Good point!

File: nat-lab/tests/utils/testing.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tomasz-grz) [reply nat-lab/tests/utils/testing.py]: `test_name, parameters = test_name.split("::", 1)[-1].split(" ")[0].split("[", 1)` actually crashes if there are no parameters

File: nat-lab/tests/utils/testing.py [hunk 8]
```diff
@@ -10,8 +12,48 @@ def unpack_optional(opt: Optional[T]) -> T:
...
-        return "".join([x if x.isalnum() else "_" for x in test_name.split(" ")[0]])
-    return test_name
+    if test_name:
+        test_parts = test_name.split("::", 1)[-1].split(" ")[0].split("[", 1)
+        if len(test_parts) > 1:
+            return (
+                format_path_string(test_parts[0]),
+                format_path_string(test_parts[1]),
+            )
+        return (format_path_string(test_parts[0]), None)
```
Comment (jjanowsk) [inline nat-lab/tests/utils/testing.py:29]: Last extra nit, feel free to ignore it. I would return either None when PYTEST_CURRENT_TEST is empty for some reason. If it's not I would return tuple of strings, not tuple of optionals. In case parameters are not present it should be an empty string. Then in `get_current_test_log_path` you wouldn't have to test if the `test_parameters` is None or not. It would be empty string so you can always concatenate it :).

File: nat-lab/tests/utils/testing.py [hunk 9]
```diff
@@ -10,8 +12,48 @@ def unpack_optional(opt: Optional[T]) -> T:
...
+    print(datetime.now(), "PYTEST_CURRENT_TEST is None")
+    return (None, None)
+
+
+def format_path_string(input_str: str) -> str:
```
Comment (jjanowsk) [inline nat-lab/tests/utils/testing.py:34]: This is now "private" to this file. We seem to have a practice of prepending those with _

**Review by jjanowsk — APPROVED** [review]
+1

File: nat-lab/tests/utils/testing.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jjanowsk) [reply nat-lab/tests/utils/testing.py]: This doesn't crash but I wouldn't use it ;)
```
 test_name, parameters, *_ = test_name.split("::", 1)[-1].split(" ")[0].split("[", 1) + [""]
```

File: nat-lab/tests/utils/testing.py (hunk 8 above)
Comment (tomasz-grz) [reply nat-lab/tests/utils/testing.py]: Yeah I thought about it, and agree, but then instead of `test_name, test_parameters = get_current_test_case_and_parameters()` directly, we would have to unpack the optional tuple first.

If we don't check `test_parameters` at all we will end up with a trailing "_", unless we prepend it in `get_current_test_case_and_parameters()` which I also don't prefer.

**Review by gytsto — APPROVED** [review]
